geomesa-spark/geomesa-spark-jts/src/main/scala/org/locationtech/geomesa/spark/jts/rules/GeometryLiteral.scala

```diff
@@ -8,24 +8,98 @@
 
 package org.locationtech.geomesa.spark.jts.rules
 
-import org.locationtech.jts.geom.Geometry
 import org.apache.spark.sql.catalyst.InternalRow
 import org.apache.spark.sql.catalyst.expressions.LeafExpression
 import org.apache.spark.sql.catalyst.expressions.codegen.CodegenFallback
-import org.apache.spark.sql.jts.JTSTypes.GeometryTypeInstance
+import org.apache.spark.sql.jts.JTSTypes._
 import org.apache.spark.sql.types.DataType
+import org.locationtech.jts.geom._
 
 /**
  * Catalyst AST expression used during rule rewriting to extract geometry literal values
  * from Catalyst memory and keep a copy in JVM heap space for subsequent use in rule evaluation.
  */
-case class GeometryLiteral(repr: InternalRow, geom: Geometry) extends LeafExpression  with CodegenFallback {
-
+abstract class GeometryLiteral extends LeafExpression with CodegenFallback {
+  def geom: Geometry
+  def repr: Any
   override def foldable: Boolean = true
-
   override def nullable: Boolean = true
-
   override def eval(input: InternalRow): Any = repr
+}
+
+object GeometryLiteral {
+
+  def unapply(g: GeometryLiteral): Option[(Any, Geometry)] = Some(g.repr, g.geom)
+
+  case class PointLiteral(geom: Point, repr: InternalRow) extends GeometryLiteral {
+    override def dataType: DataType = PointTypeInstance
+  }
+
+  object PointLiteral {
+    def apply(repr: InternalRow): PointLiteral = PointLiteral(PointTypeInstance.deserialize(repr), repr)
```

Comment:
Can case classes have secondary constructors?  (I know regular classes can.)

Reply:
hmm not sure. i think having multiple applys is generally the pattern i've seen.